### ISSUE TITLE
Joblib and Ray on K8 autoscaler don't seem to work well together when taining ML jobs

### ISSUE THREAD

**Issue by vecorro** [ray-project/ray#15500]
### What is the problem?

- Ray Server (deployed on Kubernetes) details: 'python_version': '3.7.7', 'ray_version': '2.0.0.dev0', 'ray_commit': 'bdda73e2dde0fb88f3a11a5b79071f445ec05a34', 'protocol_version': '2021-04-09'
- The Ray cluster has 1 head and 4 workers, each with 2 CPUs and 24GB of RAM
- ML library: xgboost 1.3.3 (as provided in the ray-ml Docker image)
- I'm training the Higgs dataset which has 11M examples (rows) and requires around 8GB of space.

I have made several tests on Ray deployed on K8 and the training task (using joblib) breaks when I load more than 7M rows from the dataset. Running Ray locally on my laptop completes the training task in 131 min but issues several error messages about the auto-scaler.  

Aside from the error, the Ray Dashboard shows that Ray does not make use of all the workers available. Even when at the running time, the joblib task says that will use 8 worker tasks (2CPUs x (1 master + 3 workers) perhaps), this is how it behaves most of the time:

![ray-dashboard](https://user-images.githubusercontent.com/49245780/116003583-323d8e00-a5c4-11eb-9578-980b1ee375c5.png)

**Only the master node and 1 worker heavily used and the other 2 workers doing almost nothing.**

I'm adding the Jupyter notebooks I'm using for full reproduction. The trigger will also see the error messages in there.

[scripts.zip](https://github.com/ray-project/ray/files/6372605/scripts.zip)

The Higgs dataset can be downloaded from here: https://archive.ics.uci.edu/ml/machine-learning-databases/00280/HIGGS.csv.gz

Here the error
```

Exception in thread Thread-14:
Traceback (most recent call last):
  File "/Users/kike/opt/miniconda3/envs/ray-05a34/lib/python3.7/threading.py", line 926, in _bootstrap_inner
    self.run()
  File "/Users/kike/opt/miniconda3/envs/ray-05a34/lib/python3.7/site-packages/ray/util/multiprocessing/pool.py", line 74, in run
    [ready_id], unready = ray.wait(unready, num_returns=1)
  File "/Users/kike/opt/miniconda3/envs/ray-05a34/lib/python3.7/site-packages/ray/_private/client_mode_hook.py", line 61, in wrapper
    return getattr(ray, func.__name__)(*args, **kwargs)
  File "/Users/kike/opt/miniconda3/envs/ray-05a34/lib/python3.7/site-packages/ray/util/client/api.py", line 54, in wait
    return self.worker.wait(*args, **kwargs)
  File "/Users/kike/opt/miniconda3/envs/ray-05a34/lib/python3.7/site-packages/ray/util/client/worker.py", line 281, in wait
    resp = self.server.WaitObject(req, metadata=self.metadata)
  File "/Users/kike/opt/miniconda3/envs/ray-05a34/lib/python3.7/site-packages/grpc/_channel.py", line 946, in __call__
    return _end_unary_response_blocking(state, call, False, None)
  File "/Users/kike/opt/miniconda3/envs/ray-05a34/lib/python3.7/site-packages/grpc/_channel.py", line 849, in _end_unary_response_blocking
    raise _InactiveRpcError(state)
grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:
	status = StatusCode.UNAVAILABLE
	details = "Operation timed out"
	debug_error_string = "{"created":"@1619322534.916343000","description":"Error received from peer ipv4:10.186.148.86:10001","file":"src/core/lib/surface/call.cc","file_line":1068,"grpc_message":"Operation timed out","grpc_status":14}"
>
Exception in thread Thread-15:
Traceback (most recent call last):
  File "/Users/kike/opt/miniconda3/envs/ray-05a34/lib/python3.7/threading.py", line 926, in _bootstrap_inner
    self.run()
  File "/Users/kike/opt/miniconda3/envs/ray-05a34/lib/python3.7/site-packages/ray/util/multiprocessing/pool.py", line 74, in run
    [ready_id], unready = ray.wait(unready, num_returns=1)
  File "/Users/kike/opt/miniconda3/envs/ray-05a34/lib/python3.7/site-packages/ray/_private/client_mode_hook.py", line 61, in wrapper
    return getattr(ray, func.__name__)(*args, **kwargs)
  File "/Users/kike/opt/miniconda3/envs/ray-05a34/lib/python3.7/site-packages/ray/util/client/api.py", line 54, in wait
    return self.worker.wait(*args, **kwargs)
  File "/Users/kike/opt/miniconda3/envs/ray-05a34/lib/python3.7/site-packages/ray/util/client/worker.py", line 281, in wait
    resp = self.server.WaitObject(req, metadata=self.metadata)
  File "/Users/kike/opt/miniconda3/envs/ray-05a34/lib/python3.7/site-packages/grpc/_channel.py", line 946, in __call__
    return _end_unary_response_blocking(state, call, False, None)
  File "/Users/kike/opt/miniconda3/envs/ray-05a34/lib/python3.7/site-packages/grpc/_channel.py", line 849, in _end_unary_response_blocking
    raise _InactiveRpcError(state)
grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:
	status = StatusCode.UNAVAILABLE
	details = "Operation timed out"
	debug_error_string = "{"created":"@1619322534.916355000","description":"Error received from peer ipv4:10.186.148.86:10001","file":"src/core/lib/surface/call.cc","file_line":1068,"grpc_message":"Operation timed out","grpc_status":14}"
>
Exception in thread Thread-16:
Traceback (most recent call last):
  File "/Users/kike/opt/miniconda3/envs/ray-05a34/lib/python3.7/threading.py", line 926, in _bootstrap_inner
    self.run()
  File "/Users/kike/opt/miniconda3/envs/ray-05a34/lib/python3.7/site-packages/ray/util/multiprocessing/pool.py", line 74, in run
    [ready_id], unready = ray.wait(unready, num_returns=1)
  File "/Users/kike/opt/miniconda3/envs/ray-05a34/lib/python3.7/site-packages/ray/_private/client_mode_hook.py", line 61, in wrapper
    return getattr(ray, func.__name__)(*args, **kwargs)
  File "/Users/kike/opt/miniconda3/envs/ray-05a34/lib/python3.7/site-packages/ray/util/client/api.py", line 54, in wait
    return self.worker.wait(*args, **kwargs)
  File "/Users/kike/opt/miniconda3/envs/ray-05a34/lib/python3.7/site-packages/ray/util/client/worker.py", line 281, in wait
    resp = self.server.WaitObject(req, metadata=self.metadata)
  File "/Users/kike/opt/miniconda3/envs/ray-05a34/lib/python3.7/site-packages/grpc/_channel.py", line 946, in __call__
    return _end_unary_response_blocking(state, call, False, None)
  File "/Users/kike/opt/miniconda3/envs/ray-05a34/lib/python3.7/site-packages/grpc/_channel.py", line 849, in _end_unary_response_blocking
    raise _InactiveRpcError(state)
grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:
	status = StatusCode.UNAVAILABLE
	details = "Operation timed out"
	debug_error_string = "{"created":"@1619322534.916366000","description":"Error received from peer ipv4:10.186.148.86:10001","file":"src/core/lib/surface/call.cc","file_line":1068,"grpc_message":"Operation timed out","grpc_status":14}"
>



---------------------------------------------------------------------------
_InactiveRpcError                         Traceback (most recent call last)
~/opt/miniconda3/envs/ray-05a34/lib/python3.7/site-packages/ray/util/client/worker.py in _call_schedule_for_task(self, task)
    307         try:
--> 308             ticket = self.server.Schedule(task, metadata=self.metadata)
    309         except grpc.RpcError as e:

~/opt/miniconda3/envs/ray-05a34/lib/python3.7/site-packages/grpc/_channel.py in __call__(self, request, timeout, metadata, credentials, wait_for_ready, compression)
    945                                       wait_for_ready, compression)
--> 946         return _end_unary_response_blocking(state, call, False, None)
    947 

~/opt/miniconda3/envs/ray-05a34/lib/python3.7/site-packages/grpc/_channel.py in _end_unary_response_blocking(state, call, with_call, deadline)
    848     else:
--> 849         raise _InactiveRpcError(state)
    850 

_InactiveRpcError: <_InactiveRpcError of RPC that terminated with:
	status = StatusCode.UNAVAILABLE
	details = "Operation timed out"
	debug_error_string = "{"created":"@1619322534.916374000","description":"Error received from peer ipv4:10.186.148.86:10001","file":"src/core/lib/surface/call.cc","file_line":1068,"grpc_message":"Operation timed out","grpc_status":14}"
>

During handling of the above exception, another exception occurred:

Error                                     Traceback (most recent call last)
~/opt/miniconda3/envs/ray-05a34/lib/python3.7/site-packages/sklearn/model_selection/_search.py in fit(self, X, y, groups, **fit_params)
    709 
--> 710             self._run_search(evaluate_candidates)
    711 

~/opt/miniconda3/envs/ray-05a34/lib/python3.7/site-packages/sklearn/model_selection/_search.py in _run_search(self, evaluate_candidates)
   1483             self.param_distributions, self.n_iter,
-> 1484             random_state=self.random_state))

~/opt/miniconda3/envs/ray-05a34/lib/python3.7/site-packages/sklearn/model_selection/_search.py in evaluate_candidates(candidate_params)
    688                                in product(candidate_params,
--> 689                                           cv.split(X, y, groups)))
    690 

~/opt/miniconda3/envs/ray-05a34/lib/python3.7/site-packages/joblib/parallel.py in __call__(self, iterable)
   1043 
-> 1044             while self.dispatch_one_batch(iterator):
   1045                 pass

~/opt/miniconda3/envs/ray-05a34/lib/python3.7/site-packages/joblib/parallel.py in dispatch_one_batch(self, iterator)
    858             else:
--> 859                 self._dispatch(tasks)
    860                 return True

~/opt/miniconda3/envs/ray-05a34/lib/python3.7/site-packages/joblib/parallel.py in _dispatch(self, batch)
    776             job_idx = len(self._jobs)
--> 777             job = self._backend.apply_async(batch, callback=cb)
    778             # A job can complete so quickly than its callback is

~/opt/miniconda3/envs/ray-05a34/lib/python3.7/site-packages/joblib/_parallel_backends.py in apply_async(self, func, callback)
    252         return self._get_pool().apply_async(
--> 253             SafeFunction(func), callback=callback)
    254 

~/opt/miniconda3/envs/ray-05a34/lib/python3.7/site-packages/ray/util/multiprocessing/pool.py in apply_async(self, func, args, kwargs, callback, error_callback)
    463         object_ref = self._run_batch(self._random_actor_index(), func,
--> 464                                      [(args, kwargs)])
    465         return AsyncResult(

~/opt/miniconda3/envs/ray-05a34/lib/python3.7/site-packages/ray/util/multiprocessing/pool.py in _run_batch(self, actor_index, func, batch)
    414         actor, count = self._actor_pool[actor_index]
--> 415         object_ref = actor.run_batch.remote(func, batch)
    416         count += 1

~/opt/miniconda3/envs/ray-05a34/lib/python3.7/site-packages/ray/util/client/common.py in remote(self, *args, **kwargs)
    255     def remote(self, *args, **kwargs):
--> 256         return return_refs(ray.call_remote(self, *args, **kwargs))
    257 

~/opt/miniconda3/envs/ray-05a34/lib/python3.7/site-packages/ray/util/client/api.py in call_remote(self, instance, *args, **kwargs)
     95         """
---> 96         return self.worker.call_remote(instance, *args, **kwargs)
     97 

~/opt/miniconda3/envs/ray-05a34/lib/python3.7/site-packages/ray/util/client/worker.py in call_remote(self, instance, *args, **kwargs)
    300             task.kwargs[k].CopyFrom(convert_to_arg(v, self._client_id))
--> 301         return self._call_schedule_for_task(task)
    302 

~/opt/miniconda3/envs/ray-05a34/lib/python3.7/site-packages/ray/util/client/worker.py in _call_schedule_for_task(self, task)
    309         except grpc.RpcError as e:
--> 310             raise decode_exception(e.details())
    311 

~/opt/miniconda3/envs/ray-05a34/lib/python3.7/site-packages/ray/util/client/worker.py in decode_exception(data)
    521 def decode_exception(data) -> Exception:
--> 522     data = base64.standard_b64decode(data)
    523     return loads_from_server(data)

~/opt/miniconda3/envs/ray-05a34/lib/python3.7/base64.py in standard_b64decode(s)
    104     """
--> 105     return b64decode(s)
    106 

~/opt/miniconda3/envs/ray-05a34/lib/python3.7/base64.py in b64decode(s, altchars, validate)
     86         raise binascii.Error('Non-base64 digit found')
---> 87     return binascii.a2b_base64(s)
     88 

Error: Invalid base64-encoded string: number of data characters (17) cannot be 1 more than a multiple of 4

During handling of the above exception, another exception occurred:

_InactiveRpcError                         Traceback (most recent call last)
~/opt/miniconda3/envs/ray-05a34/lib/python3.7/site-packages/ray/util/client/worker.py in _call_schedule_for_task(self, task)
    307         try:
--> 308             ticket = self.server.Schedule(task, metadata=self.metadata)
    309         except grpc.RpcError as e:

~/opt/miniconda3/envs/ray-05a34/lib/python3.7/site-packages/grpc/_channel.py in __call__(self, request, timeout, metadata, credentials, wait_for_ready, compression)
    945                                       wait_for_ready, compression)
--> 946         return _end_unary_response_blocking(state, call, False, None)
    947 

~/opt/miniconda3/envs/ray-05a34/lib/python3.7/site-packages/grpc/_channel.py in _end_unary_response_blocking(state, call, with_call, deadline)
    848     else:
--> 849         raise _InactiveRpcError(state)
    850 

_InactiveRpcError: <_InactiveRpcError of RPC that terminated with:
	status = StatusCode.UNAVAILABLE
	details = "failed to connect to all addresses"
	debug_error_string = "{"created":"@1619322555.066460000","description":"Failed to pick subchannel","file":"src/core/ext/filters/client_channel/client_channel.cc","file_line":5420,"referenced_errors":[{"created":"@1619322555.066454000","description":"failed to connect to all addresses","file":"src/core/ext/filters/client_channel/lb_policy/pick_first/pick_first.cc","file_line":398,"grpc_status":14}]}"
>

During handling of the above exception, another exception occurred:

Error                                     Traceback (most recent call last)
<ipython-input-4-33cc6d021ba4> in <module>
      1 with joblib.parallel_backend('ray', n_jobs=-1):
----> 2     search.fit(df.loc[:, df.columns != 'label'].values, df.label.values)

~/opt/miniconda3/envs/ray-05a34/lib/python3.7/site-packages/sklearn/model_selection/_search.py in fit(self, X, y, groups, **fit_params)
    708                 return results
    709 
--> 710             self._run_search(evaluate_candidates)
    711 
    712         # For multi-metric evaluation, store the best_index_, best_params_ and

~/opt/miniconda3/envs/ray-05a34/lib/python3.7/site-packages/joblib/parallel.py in __exit__(self, exc_type, exc_value, traceback)
    725 
    726     def __exit__(self, exc_type, exc_value, traceback):
--> 727         self._terminate_backend()
    728         self._managed_backend = False
    729 

~/opt/miniconda3/envs/ray-05a34/lib/python3.7/site-packages/joblib/parallel.py in _terminate_backend(self)
    755     def _terminate_backend(self):
    756         if self._backend is not None:
--> 757             self._backend.terminate()
    758 
    759     def _dispatch(self, batch):

~/opt/miniconda3/envs/ray-05a34/lib/python3.7/site-packages/joblib/_parallel_backends.py in terminate(self)
    474     def terminate(self):
    475         """Shutdown the process or thread pool"""
--> 476         super(MultiprocessingBackend, self).terminate()
    477         self.reset_batch_stats()
    478 

~/opt/miniconda3/envs/ray-05a34/lib/python3.7/site-packages/joblib/_parallel_backends.py in terminate(self)
    240         """Shutdown the process or thread pool"""
    241         if self._pool is not None:
--> 242             self._pool.close()
    243             self._pool.terminate()  # terminate does a join()
    244             self._pool = None

~/opt/miniconda3/envs/ray-05a34/lib/python3.7/site-packages/ray/util/multiprocessing/pool.py in close(self)
    651 
    652         for actor, _ in self._actor_pool:
--> 653             self._stop_actor(actor)
    654         self._closed = True
    655 

~/opt/miniconda3/envs/ray-05a34/lib/python3.7/site-packages/ray/util/multiprocessing/pool.py in _stop_actor(self, actor)
    399         # The deletion task will block until the actor has finished executing
    400         # all pending tasks.
--> 401         self._actor_deletion_ids.append(actor.__ray_terminate__.remote())
    402 
    403     def _new_actor_entry(self):

~/opt/miniconda3/envs/ray-05a34/lib/python3.7/site-packages/ray/util/client/common.py in remote(self, *args, **kwargs)
    254 
    255     def remote(self, *args, **kwargs):
--> 256         return return_refs(ray.call_remote(self, *args, **kwargs))
    257 
    258     def __repr__(self):

~/opt/miniconda3/envs/ray-05a34/lib/python3.7/site-packages/ray/util/client/api.py in call_remote(self, instance, *args, **kwargs)
     94             kwargs: opaque keyword arguments
     95         """
---> 96         return self.worker.call_remote(instance, *args, **kwargs)
     97 
     98     def call_release(self, id: bytes) -> None:

~/opt/miniconda3/envs/ray-05a34/lib/python3.7/site-packages/ray/util/client/worker.py in call_remote(self, instance, *args, **kwargs)
    299         for k, v in kwargs.items():
    300             task.kwargs[k].CopyFrom(convert_to_arg(v, self._client_id))
--> 301         return self._call_schedule_for_task(task)
    302 
    303     def _call_schedule_for_task(

~/opt/miniconda3/envs/ray-05a34/lib/python3.7/site-packages/ray/util/client/worker.py in _call_schedule_for_task(self, task)
    308             ticket = self.server.Schedule(task, metadata=self.metadata)
    309         except grpc.RpcError as e:
--> 310             raise decode_exception(e.details())
    311 
    312         if not ticket.valid:

~/opt/miniconda3/envs/ray-05a34/lib/python3.7/site-packages/ray/util/client/worker.py in decode_exception(data)
    520 
    521 def decode_exception(data) -> Exception:
--> 522     data = base64.standard_b64decode(data)
    523     return loads_from_server(data)

~/opt/miniconda3/envs/ray-05a34/lib/python3.7/base64.py in standard_b64decode(s)
    103     are discarded prior to the padding check.
    104     """
--> 105     return b64decode(s)
    106 
    107 

~/opt/miniconda3/envs/ray-05a34/lib/python3.7/base64.py in b64decode(s, altchars, validate)
     85     if validate and not re.fullmatch(b'[A-Za-z0-9+/]*={0,2}', s):
     86         raise binascii.Error('Non-base64 digit found')
---> 87     return binascii.a2b_base64(s)
     88 
     89 

Error: Invalid base64-encoded string: number of data characters (29) cannot be 1 more than a multiple of 4
```

### Reproduction (REQUIRED)

```
import numpy as np

from sklearn.datasets import load_digits
from sklearn.model_selection import RandomizedSearchCV
from xgboost.sklearn import XGBClassifier
import pandas as pd

import joblib
import ray.util.client_connect
from ray.util.joblib import register_ray

colnames = ["label"] + ["feature-%02d" % i for i in range(1, 29)]

df = pd.read_csv("https://archive.ics.uci.edu/ml/machine-learning-databases/00280/HIGGS.csv.gz", names=colnames, nrows=8000000)

param_space = {'n_estimators': range(100, 200, 20), 'max_depth': range(3, 13, 2), 'booster': ('gbtree', 'dart')}

model = XGBClassifier(random_state=0, objective='binary:logistic', use_label_encoder=False) 
search = RandomizedSearchCV(model, param_space, cv=2, n_iter=2, verbose=10)
df["label"] = df["label"].astype(int)

register_ray()

with joblib.parallel_backend('ray', n_jobs=-1):
    search.fit(df.loc[:, df.columns != 'label'].values, df.label.values)

```

):

If the code snippet cannot be run by itself, the issue will be closed with "needs-repro-script".

- [X] I have verified my script runs in a clean environment and reproduces the issue.
- [X] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/master/installation.html).

**Comment by vecorro** [comment]
UPDATE. I tried with Ray 1.3.0 and the workload distribution among workers remained very unevenly distributed (33% of the cluster's capacity is wasted):

2 cores per worker
![image](https://user-images.githubusercontent.com/49245780/117503608-cbff2680-af46-11eb-89d9-f3d67f0aef0a.png)

4 cores per worker
![image](https://user-images.githubusercontent.com/49245780/117508103-a32e5f80-af4d-11eb-8308-a56043e3197d.png)

**Comment by rkooo567** [comment]
cc @krfricke Can you follow up with this?

**Comment by krfricke** [comment]
Yeah, sorry I didn't see this. It's hard to debug without access to the infrastructure, but I'll take a look at the scripts

**Comment by vecorro** [comment]
Thanks, @krfricke, and @rkooo567. I know it's difficult. I tried to provide as much detail as possible in my report and I also tried to use the K8 setup provided in Ray's Git master branch. I see that you have a Helm chart. That probably would help get more homogeneous deployments on K8.

**Comment by krfricke** [comment]
Hi @vecorro, sorry for letting this rest so long. Did you make any progress yourself? We recently released Ray 1.4.0, my hopes would be that this might solve some of your issues.

I'm not familiar with the joblib backend and K8s autoscaling, but cc @DmitriGekhtman who might be able to help here.

Regarding the actualy training job, it seems you try to parallelize hyperparameter optimization for single-node XGBoost training jobs, right?

Instead of using joblib, you might want to look into Tune-Sklearn:

[Tune-Sklearn](https://github.com/ray-project/tune-sklearn/) is backend for hyperparameter search utilizing Ray Tune. [Here is an example for using it with single node XGBoost regressors](https://github.com/ray-project/tune-sklearn/blob/master/examples/xgbclassifier.py). Since Ray Tune uses a `SPREAD` placement strategy, you might be able to distribute the compute more evenly this way.

Also cc @Yard1 who is a maintainer of Tune-Sklearn.

**Comment by krfricke** [comment]
I'll close this for now, please feel free to re-open if the issue still comes up.
